sled-storage/src/manager.rs

```diff
@@ -0,0 +1,969 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! The storage manager task
+
+use std::collections::HashSet;
+
+use crate::dataset::{DatasetError, DatasetName};
+use crate::disk::{Disk, DiskError, RawDisk};
+use crate::error::Error;
+use crate::resources::StorageResources;
+use camino::Utf8PathBuf;
+use illumos_utils::zfs::{Mountpoint, Zfs};
+use illumos_utils::zpool::ZpoolName;
+use key_manager::StorageKeyRequester;
+use omicron_common::disk::DiskIdentity;
+use sled_hardware::DiskVariant;
+use slog::{error, info, o, warn, Logger};
+use tokio::sync::{mpsc, oneshot, watch};
+use tokio::time::{interval, Duration, MissedTickBehavior};
+use uuid::Uuid;
+
+// The size of the mpsc bounded channel used to communicate
+// between the `StorageHandle` and `StorageManager`.
+const QUEUE_SIZE: usize = 256;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum StorageManagerState {
+    WaitingForKeyManager,
+    QueuingDisks,
+    Normal,
+}
+
+#[derive(Debug)]
+struct NewFilesystemRequest {
+    dataset_id: Uuid,
+    dataset_name: DatasetName,
+    responder: oneshot::Sender<Result<(), Error>>,
+}
+
+#[derive(Debug)]
+enum StorageRequest {
+    AddDisk(RawDisk),
+    RemoveDisk(RawDisk),
+    DisksChanged(HashSet<RawDisk>),
+    NewFilesystem(NewFilesystemRequest),
+    KeyManagerReady,
+    /// This will always grab the latest state after any new updates, as it
+    /// serializes through the `StorageManager` task.
+    /// This serialization is particularly useful for tests.
```

Comment:
To be pedantic -- this is serialized after all prior requests?

Reply:
Pedantry added in 34fe6cce4